File: compiler/crates/react_estree/src/range.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::num::NonZeroU32;

use serde::ser::SerializeTuple;
use serde::{Deserialize, Serialize};

#[derive(Deserialize, Copy, Clone, Debug, PartialEq, PartialOrd, Hash)]
pub struct SourceRange {
    pub start: u32,
    pub end: NonZeroU32,
}

// ESTree and Babel store the `range` as `[start, end]`, so we customize
// the serialization to use a tuple representation.
impl Serialize for SourceRange {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut tuple = serializer.serialize_tuple(2)?;
        tuple.serialize_element(&self.start)?;
        tuple.serialize_element(&self.end)?;
        tuple.end()
    }
}

File: compiler/packages/react-forgive/server/src/utils/range.ts
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as t from '@babel/types';
import {type Position} from 'vscode-languageserver/node';

export type Range = [Position, Position];

export function isPositionWithinRange(
  position: Position,
  [start, end]: Range,
): boolean {
  return position.line >= start.line && position.line <= end.line;
}

export function isRangeWithinRange(aRange: Range, bRange: Range): boolean {
  const startComparison = comparePositions(aRange[0], bRange[0]);
  const endComparison = comparePositions(aRange[1], bRange[1]);
  return startComparison >= 0 && endComparison <= 0;
}

function comparePositions(a: Position, b: Position): number {
  const lineComparison = a.line - b.line;
  if (lineComparison === 0) {
    return a.character - b.character;
  } else {
    return lineComparison;
  }
}

export function sourceLocationToRange(
  loc: t.SourceLocation,
): [Position, Position] {
  return [
    {line: loc.start.line - 1, character: loc.start.column},
    {line: loc.end.line - 1, character: loc.end.column},
  ];
}
